jax/_src/cudnn/fused_attention_stableHLO.py

```diff
@@ -0,0 +1,672 @@
+
+# Create dot_product_attention_fwd_p for forward operation.
+_dot_product_attention_fwd_p = core.Primitive("dot_product_attention_fwd")
+_dot_product_attention_fwd_p.multiple_results = True
+_dot_product_attention_fwd_p.def_impl(partial(xla.apply_primitive, _dot_product_attention_fwd_p))
+_dot_product_attention_fwd_p.def_abstract_eval(_dot_product_attention_fwd_abstract)
+
+mlir.register_lowering(
+    _dot_product_attention_fwd_p,
+    _dot_product_attention_fwd_cuda_lowering,
+    platform="gpu",
+)
+
+_dot_product_attention_fwd_p_wrapper = core.Primitive("dot_product_attention_fwd_wrapper")
+_dot_product_attention_fwd_p_wrapper.multiple_results = True
+_dot_product_attention_fwd_p_wrapper.def_impl(_dot_product_attention_fwd_impl)
+_dot_product_attention_fwd_p_wrapper.def_abstract_eval(_dot_product_attention_fwd_abstract)
+
+# Create dot_product_attention_bwd_p for backward operation.
+_dot_product_attention_bwd_p = core.Primitive("dot_product_attention_bwd")
+_dot_product_attention_bwd_p.multiple_results = True
+_dot_product_attention_bwd_p.def_impl(partial(xla.apply_primitive, _dot_product_attention_bwd_p))
+_dot_product_attention_bwd_p.def_abstract_eval(_dot_product_attention_bwd_abstract)
+
+mlir.register_lowering(
+    _dot_product_attention_bwd_p,
+    _dot_product_attention_bwd_cuda_lowering,
+    platform="gpu",
+)
+
+_dot_product_attention_bwd_p_wrapper = core.Primitive("dot_product_attention_bwd_wrapper")
+_dot_product_attention_bwd_p_wrapper.multiple_results = True
+_dot_product_attention_bwd_p_wrapper.def_impl(_dot_product_attention_bwd_impl)
+_dot_product_attention_bwd_p_wrapper.def_abstract_eval(_dot_product_attention_bwd_abstract)
+
+
+batching.primitive_batchers[_dot_product_attention_fwd_p_wrapper] = _dot_product_attention_fwd_batcher
+batching.primitive_batchers[_dot_product_attention_bwd_p_wrapper] = _dot_product_attention_bwd_batcher
+
+_dot_product_attention_fwd_lower.def_partition(
+    infer_sharding_from_operands=_dot_product_attention_fwd_infer_sharding_from_operands,
+    partition=_dot_product_attention_fwd_partition)
+
+mlir.register_lowering(_dot_product_attention_fwd_p_wrapper,
+                        mlir.lower_fun(_dot_product_attention_fwd_lower, multiple_results=True))
+                        
+_dot_product_attention_bwd_lower.def_partition(
+    infer_sharding_from_operands=_dot_product_attention_bwd_infer_sharding_from_operands,
+    partition=_dot_product_attention_bwd_partition)
+
+mlir.register_lowering(_dot_product_attention_bwd_p_wrapper,
+                        mlir.lower_fun(_dot_product_attention_bwd_lower, multiple_results=True))
+
+dispatch.prim_requires_devices_during_lowering.add(_dot_product_attention_fwd_p)
+dispatch.prim_requires_devices_during_lowering.add(_dot_product_attention_fwd_p_wrapper)
+dispatch.prim_requires_devices_during_lowering.add(_dot_product_attention_bwd_p)
+dispatch.prim_requires_devices_during_lowering.add(_dot_product_attention_bwd_p_wrapper)
```

Comment:
Why are these needed?

Reply:
@nouiz mentioned there might be XLA crashes if not using wrapper primitives. I didn't fully understand and perhaps @nouiz could explain more on this?

Reply:
Right now, there should be a crash in XLA without that. It is @mattjj that gave this WAR. Sadly, we didn't had time to hunt and fix the XLA bug.